core/client/hammer/hammer.go

```diff
@@ -76,125 +97,199 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
+			bars = append(bars, w.bar)
+		}
+		pool, err := cpb.StartPool(bars...)
+		if err != nil {
+			return err
+		}
+	*/
 
-// launchWorkers gradually adds workers to the worker pool over ramp up to workers.
-func (h *Hammer) launchWorkers(ctx context.Context, jobs <-chan int, workers int, ramp time.Duration) <-chan error {
-	outputs := make(chan error)
+	if ok := c.TestTypes["batch"]; ok {
+		// Batch Write users
+		log.Printf("Batch Write")
+		args := genArgs(ctx, c.BatchWriteQPS, c.BatchWriteSize, c.BatchWriteCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for i := range workers {
+			handlers = append(handlers, workers[i].writeOp)
+		}
+		log.Printf("workers: %v", len(handlers))
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
 
-	go func() {
-		rampTicker := time.NewTicker(ramp / time.Duration(workers))
-		var wg sync.WaitGroup
-		for h.workers < workers {
-			h.workers++
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				h.worker(ctx, jobs, outputs)
-			}()
-			<-rampTicker.C
+	if ok := c.TestTypes["write"]; ok {
+		// Write users
+		log.Printf("User Write")
+		args := genArgs(ctx, c.WriteQPS, 1, c.WriteCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for i := range workers {
+			handlers = append(handlers, workers[i].writeOp)
 		}
-		rampTicker.Stop()
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
 
-		wg.Wait()
-		close(outputs)
-	}()
+	if ok := c.TestTypes["read"]; ok {
+		// Read users
+		log.Printf("User Read")
+		args := genArgs(ctx, c.ReadQPS, c.ReadPageSize, c.ReadCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for i := range workers {
+			handlers = append(handlers, workers[i].readOp)
+		}
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
+
+	if ok := c.TestTypes["audit"]; ok {
+		// History
+		log.Printf("User Audit History")
+		args := genArgs(ctx, c.HistoryQPS, c.HistoryPageSize, c.HistoryCount, c.Duration)
+		handlers := make([]ReqHandler, 0, len(workers))
+		for i := range workers {
+			handlers = append(handlers, workers[i].historyOp)
+		}
+		executeRequests(ctx, args, handlers)
+		fmt.Print("\n")
+	}
 
-	return outputs
+	return nil //pool.Stop()
 }
 
-// gen produces a channel of integers from 0 up to count-1 and then closes.
-func gen(ctx context.Context, count int) <-chan int {
-	jobs := make(chan int)
+func genArgs(ctx context.Context, qps, batch, count int, duration time.Duration) <-chan opArg {
+	inflightReqs := make(chan opArg, qps)
 	go func() {
-		defer close(jobs)
+		cctx, cancel := context.WithTimeout(ctx, duration)
+		defer cancel()
+		defer close(inflightReqs)
+
+		rateLimiter := rate.NewLimiter(rate.Limit(qps), qps)
 		for i := 0; i < count; i++ {
-			select {
-			case <-ctx.Done():
+			userIDs := make([]string, 0, batch)
+			for j := 0; j < batch; j++ {
+				userIDs = append(userIDs, fmt.Sprintf("user_%v", i*batch+j))
+			}
+			if err := rateLimiter.WaitN(cctx, batch); err != nil {
```

Comment:
Note that WaitN `... returns an error if n exceeds the Limiter's burst size`, i.e. in your case if `batch > qps`. For default ones listed in flags it works, but not for arbitrary params. If you wanted to make it work always, you would need to create the limiter like this: `rate.NewLimiter(rate.Limit(qps), batch)`.

Reply:
done